allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py

```diff
@@ -398,7 +397,7 @@ def forward(self,  # type: ignore
                 if i in best_final_states:
                     best_action_indices = best_final_states[i][0].action_history[0]
                     if target_action_sequences is not None:
-                        # Use a Tensor, not a Variable, to avoid a memory leak.
+                        # Detach to avoid a memory leak.
                         targets = target_action_sequences[i].data
```

Comment:
I was a little confused on this, here's what the migration docs say:

```
However, .data can be unsafe in some cases. Any changes on x.data wouldn’t be tracked by autograd, and the computed gradients would be incorrect if x is needed in a backward pass. A safer alternative is to use x.detach(), which also returns a Tensor that shares data with requires_grad=False, but will have its in-place changes reported by autograd if x is needed in backward.
```

to me that suggests that `detach` still retains some connection to the computational graph and might still result in a memory leak?

Reply:
I just asked about this in the pytorch slack, and I got this response:

"Variables returned by `.detach()` do not hold on to the computation graph. You can almost certainly use `.detach()` in places where you are using `.data`

"They do share a version counter with the original tensor. This doesn't require any extra memory usage. It just means that changes to the detached variable increment the (shared) version counter on  original tensor. So if you do backprop on the original tensor, and it's needed for a gradient computation, you would get an error message instead of just incorrect gradients."